sys/drv/umass.c

```diff
@@ -0,0 +1,528 @@
+
+/* Perform the read capacity command. */
+static int umass_read_capacity(device_t *dev) {
+  scsi_read_capacity_data_t cap_data;
+  scsi_read_capacity_t cap = (scsi_read_capacity_t){
+    .opcode = READ_CAPACITY,
+  };
+  int i, error = 0;
+
+  for (i = 0; i < READ_CAPACITY_MAX_ITR; i++) {
+    /* Try to issue the read capacity command. */
+    error = umass_transfer(dev, &cap_data, sizeof(scsi_read_capacity_data_t),
+                           USB_DIR_INPUT, &cap, sizeof(scsi_read_capacity_t));
+    if (!error)
+      break;
+
+    /* We've failed. Let's perform the request sense command. */
+    error = umass_request_sense(dev);
+    if (error) {
+      /* This kind of error is unacceptable. Return with error. */
+      return error;
+    }
+  }
+  if (i == READ_CAPACITY_MAX_ITR) {
+    /* We did our best. */
+    return ENODEV;
+  }
+
+  /* We assume that number of logical blocks is <= UINT32_MAX. */
+  if (cap_data.addr == UINT32_MAX)
+    return ENXIO;
+
+  /* Save number of logical blocks and block size. */
+  umass_state_t *umass = dev->state;
+  umass->nblocks = be32toh(cap_data.addr) + 1;
+  umass->block_size = be32toh(cap_data.length);
+
+  return 0;
+}
+
+/*
+ * Read/Write functions.
+ */
+
+static int umass_read(vnode_t *v, uio_t *uio) {
+  device_t *dev = devfs_node_data(v);
+  umass_state_t *umass = dev->state;
+  uint32_t block_size = umass->block_size;
+  off_t offset = uio->uio_offset;
+  int error = 0;
+
+  /* Obtain the LBA of the first block to read. */
+  uint32_t start = offset / block_size;
+
+  /* Obtain the LBA of the last block to read. */
+  uint32_t end = roundup(offset + uio->uio_resid, block_size) / block_size;
+
+  /* How many blocks is it?
+   * Note: we assume that number of blocks to read is <= UINT16_MAX. */
+  uint16_t nblocks = end - start;
+  uint32_t size = nblocks * block_size;
+
+  void *buf = kmalloc(M_DEV, size, M_WAITOK);
+  if (!buf)
+    return ENOMEM;
+
+  scsi_rw_10_t rw10 = (scsi_rw_10_t){
+    .opcode = READ_10,
+    .addr = htobe32(start),
+    .length = htobe16(nblocks),
+  };
+  error =
+    umass_transfer(dev, buf, size, USB_DIR_INPUT, &rw10, sizeof(scsi_rw_10_t));
+  if (error)
+    goto end;
+
+  /* Copy the data to the user. */
+  uio->uio_offset -= start;
+  error = uiomove_frombuf(buf, size, uio);
+
+  /* XXX: this is a cheat which keeps file offset at 0.
+   * This should be removed when a better interface is presended. */
+  uio->uio_offset = 0;
+
+end:
+  kfree(M_DEV, buf);
+  return error;
+}
+
+static int umass_write(vnode_t *v, uio_t *uio) {
+  device_t *dev = devfs_node_data(v);
+  umass_state_t *umass = dev->state;
+  uint32_t block_size = umass->block_size;
+  off_t offset = uio->uio_offset;
+  int error = 0;
+
+  /* Obtain the LBA of the first block invloved. */
+  uint32_t start = offset / block_size;
+
+  /* Obtain the LBA of the last block involved. */
+  uint32_t end = roundup(offset + uio->uio_resid, block_size) / block_size;
+
+  /* How many blocks is it?
```

Comment:
Excuse me, but who is supposed to answer that question? 
If that question is supposed to explain the value of `nblocks` then rewrite it's a sentence